---
output: pdf_document
geometry: margin=1in
mainfont: Times New Roman
---

```{r setup,warning=F, message=F, echo=F,comment="K",results='asis'}
library(pander)
library(dplyr)
library(ggplot2)
library(data.table)
library(grid)
library(stringr)
library(gridExtra)
library(knitr)
library(tidyr)
library(scales)
library(gtable)
library(stationaRy)
library(stringr)
library(cegwas)
library(lubridate)
library(cowplot)
library(plyr)

panderOptions('knitr.auto.asis', TRUE)

setwd("~/Dropbox/AndersenLab/LabFolders/Katie/projects/noaa/data/")

#Import data
load("20161103.noaa.pmd.all.RData")
pmd3mo <- processed_mapping_df_3mo %>% mutate(trait = paste0(trait, "_3mo"))
pmd1yr <- processed_mapping_df_1yr %>% mutate(trait = paste0(trait, "_1yr"))
pmd3yr <- processed_mapping_df_3yr %>% mutate(trait = paste0(trait, "_3yr"))
pmdloc <- readRDS("~/Dropbox/AndersenLab/LabFolders/Katie/projects/noaa/data/20160715.pmd.152.location.rds") %>%
  mutate(time_period = c(NA))
pmd3mo_var <- readRDS("20161104_pmd_var_3mo") %>% mutate(trait = paste0(trait, "_var_3mo"), time_period = c(NA))
pmd1yr_var <- readRDS("20161104_pmd_var_1yr") %>% mutate(trait = paste0(trait, "_var_1yr"), time_period = c(NA))
pmd3yr_var <- readRDS("20161104_pmd_var_3yr") %>% mutate(trait = paste0(trait, "_var_3yr"), time_period = c(NA))
processed_mapping_df <- rbind(pmdloc, pmd3mo, pmd1yr, pmd3yr, pmd3mo_var, pmd1yr_var, pmd3yr_var)
wi_var <- read.csv("20161104_wi_var.csv") %>% dplyr::select(-X)
wi_weather <- wi_weather %>% dplyr::select(-X)
wi_weather2 <- join(wi_weather, wi_var)

```

```{r,fig.width=7.5, fig.height=3,warning=F, message=F, echo=F,comment="K",results='asis', dpi = 300}
#PxG function
hm<-processed_mapping_df
hm<-distinct(processed_mapping_df, trait,strain,peak_id, .keep_all = T)
hm$allele <- factor(hm$allele,
                          levels = c(-1,1),
                          labels = c("REF", "ALT"))

gwasPxG <- function(trt,specific_peak){
  hm %>%
    filter(trait==trt,peak_id==specific_peak,!is.na(allele))%>%
    ggplot(.)+
    aes(x=allele,y = value,fill=as.factor(allele))+
    geom_boxplot(outlier.shape=NA,size =.5,color="gray52")+
    geom_point(size = 1, alpha = .8,position=position_jitter(w=.4,  h=.025),na.rm=TRUE)+
    theme_bw()+
    theme(axis.text.x = element_text(size=9, color="black"),
          axis.text.y = element_text(size=9, color="black"),
          axis.title = element_text(size=9, color="black", vjust = 1, face = "bold"),
          strip.text.x = element_text(size=9, color="black"),
          strip.text.y = element_text(size=9, color="black"),
          plot.title = element_text(size=9, vjust=1),
          legend.title = element_text(size=9),
          panel.border = element_rect(size=1, colour = "black"),
          plot.margin = unit(c(.05,.05,.05,.05), "cm"),
          legend.position = "none")+
          scale_y_continuous(breaks= pretty_breaks())+
    labs( x = "Genotype",y="Value")+
    scale_fill_manual( values = c("darkgray", "burlywood2", "darkolivegreen","black"))
}

#LD plot function (adapted from cegwas::plot_peak_ld)
plot_peak_ld.kt <- function (plot_df, trait = NULL) {
  plot_df <- plot_df %>% dplyr::select(-time_period)
  if (is.null(trait)) {
    snp_df <- plot_df %>% na.omit()
  } else {
    snp_df <- dplyr::filter(plot_df, trait == trait) %>% 
              na.omit()
  }
  ld_snps <- dplyr::filter(snps, CHROM %in% snp_df$CHROM, POS %in% snp_df$POS)
  ld_snps <- data.frame(snp_id = paste(ld_snps$CHROM, ld_snps$POS, sep = "_"),
                        data.frame(ld_snps)[, 5:ncol(ld_snps)])
  sn <- list()
  for (i in 1:nrow(ld_snps)) {
    sn[[i]] <- genetics::genotype(as.character(gsub(1, "T/T", 
                                                    gsub(-1, "A/A", ld_snps[i,4:ncol(ld_snps)]))))
  }
  test <- data.frame(sn)
  colnames(test) <- (ld_snps$snp_id)
  if (ncol(test) == 1) {
    #print("Only one significant SNP, not calculating LD")
  }
  else {
    ldcalc <- t(genetics::LD(test)[[3]])
    diag(ldcalc) <- 1
    LDs <- tbl_df(data.frame(ldcalc) %>% 
            dplyr::add_rownames(var = "SNP1")) %>%
            tidyr::gather(SNP2, Dprime, -SNP1) %>% 
            dplyr::arrange(SNP1) %>% 
            tidyr::separate(SNP1, sep = "_", into = c("CHROM1", "POS1"), remove = F) %>% 
            dplyr::arrange(CHROM1, as.numeric(POS1))
    ldplot <- ggplot2::ggplot(LDs) + 
              ggplot2::aes(x = factor(SNP1, levels = SNP1, ordered = T), 
                           y = factor(SNP2, levels = SNP1, ordered = T)) + 
              ggplot2::geom_tile(ggplot2::aes(fill = Dprime)) + 
              ggplot2::geom_text(ggplot2::aes(label = signif(Dprime, 3)), size = 6) + 
              ggplot2::theme(axis.text.x = ggplot2::element_text(size = 8, color = "black"), 
                             axis.text.y = ggplot2::element_text(size = 8, color = "black", 
                                                                 angle = -90, hjust = 0.5), 
                             axis.title.x = ggplot2::element_text(size = 0, color = "black", 
                                                                  vjust = -0.3), 
                             axis.title.y = ggplot2::element_text(size = 0, color = "black"), 
                             legend.position = "none") + 
      scale_x_discrete(labels = function(x) {
        gsub("_", ":", x)
      }, expand = c(0, 0)) + scale_y_discrete(labels = function(x) {
        gsub("_", ":", x)
      }, expand = c(0, 0)) + scale_fill_continuous(high = "#FF0000", low = "white", 
                                                   na.value = "white")
    ldplot <- cowplot::ggdraw(cowplot::switch_axis_position(ldplot, "y"))
    return(ldplot)
  }
}

#Make manhattan plots (adapted from cegwas::manplot)
traits <- c("elevation", "rh_avg_3mo", "rh_max_3mo", "rh_var_3mo", "wd_var_3mo", "ws_max_3mo", "ceil_hgt_max_1yr", "rh_avg_1yr", "rh_max_1yr", "rh_var_1yr", "temp_var_1yr", "wd_max_1yr", "wd_var_1yr", 'ws_var_1yr', "dew_point_max_3yr", "dew_point_var_3yr", "rh_avg_3yr", "rh_max_3yr", 'temp_avg_3yr', 'temp_min_3yr', 'temp_var_3yr', 'wd_max_3yr', 'wd_var_3yr', 'ws_var_3yr')

for(i in traits) {
  plot_title <- paste0("plot_", i)
  selection <- processed_mapping_df %>% filter(trait == i)
  max <- max(selection$log10p)
  assign(plot_title, processed_mapping_df %>% 
           dplyr::filter(trait == i) %>% 
           dplyr::distinct(marker, .keep_all = T) %>% 
           ggplot2::ggplot(.) + 
           ggplot2::aes(x = POS/1e+06, y = log10p, fill = BF) + 
           ggplot2::geom_rect(mapping=aes(xmin=startPOS/1e6, xmax=endPOS/1e6, ymin=0, 
                                          ymax= Inf),fill="thistle1", alpha=1) +
           ggplot2::geom_point(aes(color = ifelse(log10p > BF, 'red', 'black')), size = 1) +
           ggplot2::facet_grid(. ~ CHROM, scales = "free_x", space = "free_x") + 
           ggplot2::geom_hline(ggplot2::aes(yintercept = BF), color = "grey60", 
                               linetype = "dashed") + 
           ggplot2::theme(strip.background = element_blank(),
                          strip.text.x = element_text(size = 9, face = "bold"),
                          panel.background = element_rect(fill = "white"),
                          panel.border = element_rect(color="black", size=0.5,
                                                      linetype="solid", fill=NA),
                          panel.margin = unit(.6, "lines"),
                          panel.background = element_blank(),
                          axis.ticks = element_line(colour = "black"),
                          axis.text.y = element_text(colour = "black", size = 9),
                          axis.text.x = element_text(colour = "black"),
                          axis.title = element_text(colour = "black", size = 9, face = "bold"),
                          axis.line.x = element_line(size = 0),
                          axis.line.y = element_line(size = 0),
                          legend.position=('none'))+
          ggplot2::labs(x = "Genomic Position (Mb)", 
                        y = expression(bold(-log["10"](p))), colour = "red") +
          ggplot2::scale_color_identity()+
          ggplot2::scale_fill_continuous(name = i) +
          scale_y_continuous(breaks= pretty_breaks(),expand=c(0,0),limits=c(0,max+.075*max),
                             labels = function(x) format(x, width = 4))
         
  )
  #Create title
  if(i == "elevation") { title <- "Elevation (m)"; type <- ""; time <- ""} else 
    if(i == "ceil_hgt_max_1yr") { title <- "Cloud Height (m)"; time <- "1yr"; type <- "max" } else 
      if(i == "dew_point_max_3yr") { title <- "Dew Point (C)"; type <- "max"; time <- "3yr"} else 
        if(i == "dew_point_var_3yr") { title <- "Dew Point"; type <- "var"; time <- "3yr"} else {
    list <- strsplit(i, "_")[[1]]
    title <- list[1]
    type <- list[2]
    time <- list[3]
  }
  if(title == "rh") { title <- "Relative Humidity (%)"} else
    if(title == "temp") { title <- "Temperature (C)"} else
      if(title == "ws") {title <- "Wind Speed (m/s)"} else
        if(title == "wd") {title <- "Wind Direction (Degrees)"}
  if(time == "3mo") { time <- "3 months"} else
    if(time == "1yr") { time <- "1 year"} else
      if(time == "3yr") { time <- "3 years"}
  if(type == "avg") { type <- "Average Daily"} else
    if(type == "min") { type <- "Minimum Daily"} else
      if(type == "max") { type <- "Maximum Daily"} else
        if(type == "var") { 
          type <- "Variance of"
          if(title == "Relative Humidity (%)") {title <- "Relative Humidity"} else
            if(title == "Temperature (C)") { title <- "Temperature"} else
              if(title == "Wind Speed (m/s)") {title <- "Wind Speed"} else
                if(title == "Wind Direction (Degrees)") {title <- "Wind Direction"}
          }
  if(time == "") { title <- paste0(type, " ", title)} else {
  title <- paste0(type, " ", title, " [", time, "]") }

  #plot manhattan plot
  print(plot_grid(get(plot_title) + ggtitle(title)) + draw_label("A", x = 0.03, y = 0.83, fontface = "bold"))
  #manplot <- get(plot_title) + ggtitle(title)
  cat("\n\n\n") #new line whitespace in between graphs

  plist<-vector()
  for (p in unique(selection$peak_id)){
    plist<-c(plist,p)
  }
  plist<-sort(plist)

  #plot pxg plots in one row
  box_list <- lapply(c(plist),FUN=function(x){gwasPxG(i,x)})
  blank_plot <- NULL
  #pxg <- (do.call("grid.arrange", c(box_list, ncol=length(plist)))) 
  if(length(plist) == 1) { print(plot_grid(box_list[[1]], blank_plot, nrow = 1, labels = "B", hjust = -0.1)) } else
  if(length(plist) == 2) { print(plot_grid(box_list[[1]], box_list[[2]], blank_plot, nrow = 1, labels = "B", hjust = -0.1)) } else
  if(length(plist) == 3) { print(plot_grid(box_list[[1]], box_list[[2]], box_list[[3]], nrow = 1, labels = "B", hjust = -0.1)) } else
  if(length(plist) == 4) { print(plot_grid(box_list[[1]], box_list[[2]], box_list[[3]], box_list[[4]], nrow = 1, labels = "B", hjust = -0.1)) } else
  if(length(plist) == 5) { print(plot_grid(box_list[[1]], box_list[[2]], box_list[[3]], box_list[[4]], box_list[[5]], nrow = 1, labels = "B", hjust = -0.1)) }

  cat("\n\n\n")
  
  #If more than 1 QTL, plot LD plot
  blank_plot <- NULL
  if(length(unique(selection$startPOS)[!is.na(unique(selection$startPOS))]) > 1) {
    print(plot_grid(plot_peak_ld.kt(selection), blank_plot, nrow = 1, labels = "C", hjust = -0.1))
  } 
  
  #cat("\n\n\\pagebreak\n")
  ggsave(filename = paste0("~/Dropbox/AndersenLab/LabFolders/Katie/projects/noaa/figures/figureS", count))

}

```





